### DISCUSSION TITLE
How to modifiy glsl shaders to sample multisampled textures?

### DISCUSSION THREAD

**Discussion by ifsheldon** [gfx-rs/wgpu#1536] [Q&A]
Hi guys! I finished my toy 3D renderer and then I try to do simple multisampling, so I am wondering how I can modify my shader code to enable it to sample multi-sampled textures.

My `glsl` fragment shader code contains uniforms like:
```glsl
#version 450
layout(set=0, binding = 0) uniform texture2D front_face_tex;
layout(set=0, binding = 1) uniform sampler front_face_sampler;
```
and in the shader's main, I use it like
```glsl
vec3 color= texture(sampler2D(front_face_tex, front_face_sampler), texCoord).rgb;
```
The above code works fine when multisampling is disabled, but when I enabled multisampling, it gave me runtime errors
```shell
Caused by:
In Device::create_render_pipeline
  note: label = `Render Pipeline`
error matching FRAGMENT shader requirements against the pipeline
shader global ResourceBinding { group: 0, binding: 0 } is not available in the layout pipeline layout
texture class Sampled { kind: Float, multi: true } doesn't match the shader Sampled { kind: Float, multi: false }
```
I think this error indicates that I got the configurations on CPU side correct, but I need to somehow modify my shader code.

When I configure `BindGroupLayout`, I see the field `multisampled` in `BindingType::Texture` has documentation that says "If this is true, the texture must be read from shaders with texture1DMS, texture2DMS, or texture3DMS, depending on dimension.". 
So, I tried adding `MS` to different places, but `shaderc` always complains.

For the CPU-side code, basically I just modified the following two structs and their fields:
* `MultisampleState.count`
* `TextureDescriptor.sample_count`

Do you have any suggestions? Thank you!

**Comment by kvark** [comment]
You can't "sample" from a multi-sampled texture (in either of the APIs). You can only load/fetch samples from it:
```
  gvec4 texelFetch(gsampler2DMS sampler, ivec2 P, int sample)
  gvec4 texelFetch(gsampler2DMSArray sampler, ivec3 P, int sample)
```

**Reply by ifsheldon** [reply]
OK.........but it seems now I need to manually interpolate texture values. 

Now I am actually doing multipass rendering by rendering intermediate results into a texture. The intermediate results and the final image have the same resolution, so I wonder if I can improve the quality of intermediate results and the final image just by enabling multisampling for all passes. Based on this context, do you have any suggestions or workarounds? Thank you!

**Reply by kvark** [reply]
> OK.........but it seems now I need to manually interpolate texture values

Yes, that's the point. HW can't do interpolation across multiple AA samples. You need to either manually load the values from each, or use an MSAA resolve first into a regular texture.

Please take a look at the msaa-line example. You probably want to do it similarly: rendering to MSAA surface, but then resolving into a regular texture, or to the swapchain frame.

**Reply by ifsheldon** [reply]
Yeah, I had taken a glimpse on that before I asked this and my trials, but it's a bit confusing in that I cannot see method call dependencies clearly only from the MSAA source code. It seems all examples have a "framework" to get preparations done and what are in each example are just implemented "hooks". Anyway, I will take a closer look.

Although it's not relevant to this discussion, I hope the examples can be standalone, having redundant code that does the preparations and has less abstractions, or at least, in each example, there is a documentation explaining conceptually the call sequence of the implemented trait methods.
